### DISCUSSION TITLE
GPU encoding quality

### DISCUSSION THREAD

**Discussion by nums** [OvenMediaLabs/OvenMediaEngine#1296] [Help]
Hello, 

I made some test with GPU encoding, and it seems render quality is lower than CPU encoding.

This is some screen captures (click on it to see in real format) to show the difference between CPU and GPU encoding, I made the test with this configuration : 
```
<Video>
  <Name>720p</Name>
  <Codec>h264</Codec>
  <Bitrate>2024000</Bitrate>
  <Framerate>30</Framerate>
  <Width>1280</Width>
  <Height>720</Height>
  <Preset>slower</Preset>
</Video>
```
I have checked if the GPU was full during and it is not the case. 

**CPU:**
![codec-compare-cpu-720p](https://github.com/AirenSoft/OvenMediaEngine/assets/274268/2318bb5a-0205-47b6-9d49-e330195934ef)

**GPU:**
![codec-compare-nvenc-720p](https://github.com/AirenSoft/OvenMediaEngine/assets/274268/1e0f2243-1914-4eef-ad70-23778704c0ab)

The main quality difference appear with screencasting :

**CPU:**
![codec-compare-screen-cpu-720p](https://github.com/AirenSoft/OvenMediaEngine/assets/274268/c639b2f3-094a-4c60-a2a5-d11c8c92c882)

**GPU:**
![codec-compare-screen-nvenc-720p](https://github.com/AirenSoft/OvenMediaEngine/assets/274268/68d29be9-e117-4d09-945f-f23648dad9d0)

Letters are very aliased on GPU, texts are not really readable

Is there a way to improve this ?

**Comment by Keukhan** [comment]
@nums 

In order to be compatible with the WebRTC protocol and use system resources as little as possible, the baseline profile, no bframe, and KeyframeInterval are set to 1 second as default settings, so the quality may look low.

First, try changing the KeyFrameInterval and Profile.  
If you use LL-HLS instead of WebRTC for playback, try setting the BFrames value. 

Guide : https://airensoft.gitbook.io/ovenmediaengine/transcoding#encodes

If you want to change detailed settings, please edit the file in the /projects/transcoder/encoder/encoder_avc_nv.cpp path.

Thanks.

**Comment by nums** [comment]
After some testing and recompliation with other presets (like level or changing tune ull by hq) there is no quality update.
Maybe the quality issue is in the scaling?

I saw that scale_npp is preferred to scale_cuda : https://stackoverflow.com/questions/55687189/how-to-use-gpu-to-accelerate-the-processing-speed-of-ffmpeg-filter/55747785#55747785

I don't know if you ever test scale_npp but I can test on my side.

**Comment by Keukhan** [comment]
@nums 
How is the original quality? Comparing encoding option to Passthrough, is there a big difference?

**Comment by nums** [comment]
This is a screen of the original quality, but you can see the difference between 720p transcoding with cpu and gpu in my first post : the readability is not the same.

![image - 2023-07-07T091147 921](https://github.com/AirenSoft/OvenMediaEngine/assets/274268/3b60c2f3-3aca-48f8-9098-5643d478f59c)

I haven't yet managed to use scale_npp with OvenMedia, I have an issue but I will do some test with FFMPEG cli and compare scaling 1080p to 720p video with diffretents parameters.

**Comment by nums** [comment]
I've made test with the replacing of `"hwupload_cuda,scale_cuda=w=%d:h=%d,hwdownload"`, to `"scale=%dx%d:flags=bilinear",`
and the quality looks better. GPU is still in use, but it seems that the scaling processing workflow CPU (hwupload_cuda) => GPU (hwdownload) => CPU is degrading the frame.

![image - 2023-07-07T150101 812 (1)](https://github.com/AirenSoft/OvenMediaEngine/assets/274268/6ccc43dc-b91e-4c7c-8632-6903190f64d9)

What do you think about a setting like `<Rescaler>CPU</Rescaler>` in the Server.xml to be able to force rescaler to not use GPU ?

**Reply by Keukhan** [reply]
@nums 

Thank you for your valuable information.. 
CPU base scaler between the GPU decoder and encoder does increase the overhead of copying between host memory and GPU memory.  So, forcibly using the scaler as S/W requires consideration.  I will find out why scale_cuda is of poor quality and how to improve it.

I'm busy making money these days, so I can't test it right now, but I'll try it soon. :)
If you have a good approach in the meantime, please share. It will be of great help to the development of OME.

**Comment by nums** [comment]
Thanks @Keukhan for your answer.

On my side, I can use CPU for the rescaler and continue to use GPU for encoding and and it's quite satisfying.
Are you interested a PR with this management ?

I don't know where is the most pertinent: 

I can have a new node at the same level of : 
`<HardwareAcceleration>true</HardwareAcceleration>`
with this kind of node: 
`<Recaler>cpu</Recaler>`

or we can do this for each encoding like that: 
```
<Video>
  <Codec>h264</Codec>
  <Width>1920</Width>
  <Height>1080</Height>
  <Bitrate>5000000</Bitrate>
  <Framerate>30</Framerate>
  <Recaler>cpu</Recaler>
</Video>
```

Tell me if it's relevant and I'll do the PR ;)

**Comment by nums** [comment]
Hi @Keukhan,

I see you are working on the GPU Docker image.
What do you think of my proposal ? Can I make a PR or have you got a other idea ?

**Reply by ciroiriarte** [reply]
out of curiosity, on similar outcomes with CPU and GPU, what's the efficiency you see in resource usage?.

I'm trying to understand if adding GPU in the mix allows capacity to increase in fractions or multipliers for concurrent encoding.

**Reply by nums** [reply]
Hello @ciroiriarte ,

It depends on the type of GPU and the number of CPUs used. For my part, using an Nvidia Pro GPU from GCP or Scaleway allows me to activate between 5 to 10 streams each with 3 transcodes (bypass, 1080p, 720p, 360p), at the CPU level to have 5 live transcoded, I need 2 to 3 threads per encoding, or at least 24 CPUs (with the rest of the processes). 
Using the GPU for encoding allows me to use the CPU for other tasks on the same machine (but this is my use case).

Furthermore, when I modified the "scale" process I found the quality of the GPU rendering better than the CPU.

**Reply by Keukhan** [reply]
@nums
Thank you always for your good comments.
I have been so busy lately that I couldn't answer. Please understanding.

I've been working on multi-GPU support for a few days now. Among them, I was thinking about the function to set hardware activation of Scaler. This is a flexible setup, but I thought it was getting complicated. So, I plan to limit the scaler to be automatically adopted depending on the encoder type.
If you do a PR, I'll review it and merge it if it's good for usability.

Thank you for your contribution to the development of OME.

**Reply by nums** [reply]
@Keukhan 
No worries and thanks for the work done for multi-gpu support. For my part I will also try to make ovenmedia compatible with the netint VPU card (https://github.com/AirenSoft/OvenMediaEngine/discussions/1397)

I'm also going to re-investigate this scaling thing because I hadn't tested everything, it is also possible to add the interp_algo parameter which has several rescaling algorithms. The answer may be here, and it may be the algorithm that we will need to be able to configure.

I will move forward on these subjects in November and I will keep you informed.

**Reply by ciroiriarte** [reply]
> Hello @ciroiriarte ,
> 
> It depends on the type of GPU and the number of CPUs used. For my part, using an Nvidia Pro GPU from GCP or Scaleway allows me to activate between 5 to 10 streams each with 3 transcodes (bypass, 1080p, 720p, 360p), at the CPU level to have 5 live transcoded, I need 2 to 3 threads per encoding, or at least 24 CPUs (with the rest of the processes). 
> Using the GPU for encoding allows me to use the CPU for other tasks on the same machine (but this is my use case).

So, with CPU offloading you still need 24 CPU cores. How would the CPU core count look like without GPU offloading?.

> Furthermore, when I modified the "scale" process I found the quality of the GPU rendering better than the CPU.

Hope you publish the setup procedure :)

**Reply by nums** [reply]
> So, with CPU offloading you still need 24 CPU cores. How would the CPU core count look like without GPU offloading?.

Oh no sorry, I may have expressed myself incorrectly, 24 processors are necessary when I am not using the GPU.

**Reply by ciroiriarte** [reply]
oh, ok. And how does that number change when introducing 1 GPU for your 10 streams with 3 transcodes each?.

I find interesting your comment about getting better quality, while I always hear that best quality comes from CPU. 
Example: https://www.tomshardware.com/news/amd-intel-nvidia-video-encoding-performance-quality-tested#:~:text=GPU%20video%20encoding%20is%20generally,files%20and%20higher%20average%20bitrates.

Looking forward for the details :)
